apps/staking/app/register/[nodeId]/Registration.tsx

```diff
@@ -0,0 +1,572 @@
+        }
+      }
+
+      let isExpress = userPrefMode === REG_MODE.EXPRESS;
+
+      if (qpMode === REG_MODE.EXPRESS) {
+        isExpress = true;
+      } else if (qpMode === REG_MODE.GUIDED) {
+        isExpress = false;
+      }
+
+      let queryParamStartTab = null;
+
+      if (qpNodeType === NODE_TYPE.SOLO) {
+        queryParamStartTab = REG_TAB.SUBMIT_SOLO;
+      } else if (qpNodeType === NODE_TYPE.MULTI) {
+        if (data?.stakeAmount === undefined) {
+          queryParamStartTab = REG_TAB.STAKE_AMOUNT;
+        } else if (data?.operatorFee === undefined) {
+          queryParamStartTab = REG_TAB.OPERATOR_FEE;
+        } else if (isExpress) {
+          queryParamStartTab = REG_TAB.SUBMIT_MULTI;
+        } else if (data?.rewardsAddress === undefined) {
+          queryParamStartTab = REG_TAB.REWARDS_ADDRESS;
+        } else if (data?.autoActivate === undefined) {
+          queryParamStartTab = REG_TAB.AUTO_ACTIVATE;
+          // TODO: implement the reserve slots tab
+          // } else if (data?.reservedContributors === undefined) {
+          //   queryParamStartTab = REG_TAB.RESERVE_SLOTS;
+        } else {
+          queryParamStartTab = REG_TAB.SUBMIT_MULTI;
+        }
+      }
+
+      if (queryParamStartTab !== null) {
+        setTabHistory([REG_TAB.START]);
+      }
+
+      return {
+        queryParamFields: data,
+        queryParamMode: qpMode,
+        queryParamNodeType: qpNodeType,
+        queryParamStartTab,
+      };
+    }, []);
+
+  const alreadyRegisteredMulti =
+    contract !== null &&
+    (contract.status === CONTRIBUTION_CONTRACT_STATUS.WaitForOperatorContrib ||
+      contract.status === CONTRIBUTION_CONTRACT_STATUS.OpenForPublicContrib);
+
+  const [tab, setTab] = useState<REG_TAB>(
+    addedBlsKeys?.has(blsKey)
+      ? REG_TAB.ALREADY_REGISTERED_RUNNING
+      : alreadyRegisteredMulti
+        ? REG_TAB.ALREADY_REGISTERED_MULTI
+        : queryParamStartTab ?? REG_TAB.START
+  );
+  const [nodeType, setNodeType] = useState<NODE_TYPE>(
+    alreadyRegisteredMulti ? NODE_TYPE.MULTI : queryParamNodeType ?? NODE_TYPE.SOLO
+  );
+  const [mode, setMode] = useState<REG_MODE>(queryParamMode ?? userPrefMode);
+
+  function setStartData(data: z.infer<typeof RegistrationStartFormSchema>) {
+    setMode(data.mode);
+    pushQueryParam(REGISTRATION_QUERY_PARAM.MODE, data.mode);
+
+    setNodeType(data.nodeType);
+    pushQueryParam(REGISTRATION_QUERY_PARAM.IS_MULTI, data.nodeType === NODE_TYPE.MULTI);
+
+    changeTab(data.nodeType === NODE_TYPE.SOLO ? REG_TAB.SUBMIT_SOLO : REG_TAB.STAKE_AMOUNT);
+  }
+
+  const changeTab = (targetTab: REG_TAB) => {
+    setTabHistory((prev) => [...prev, tab]);
+    setTab(targetTab);
+    setBackButtonClickCallback(null);
+  };
+
+  const handleTabBackButtonClick = () => {
+    if (backButtonClickCallback) {
+      backButtonClickCallback();
+    }
+    setTab(tabHistory.at(-1) ?? REG_TAB.START);
+    setTabHistory((prev) => prev.slice(0, -1));
```

Comment:
tab history could be undefined. I realise you handle it above but feels iffy

Reply:
`tabHistory` cant be undefined it initializes as an empty array

Reply:
My concern was more with Arrray.slice() but seems it resolve to an empty array if you try and do -1 on an empty one so that is fine. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#end

Resolving.